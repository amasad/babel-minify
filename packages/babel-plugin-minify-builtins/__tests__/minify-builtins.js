jest.autoMockOff();

const babel = require("babel-core");
const unpad = require("../../../utils/unpad");
const plugin = require("../src/index");

function transform(code) {
  return babel.transform(code, {
    plugins: [plugin]
  }).code;
}

describe("minify-builtins", () => {
  it("should minify standard built in methods", () => {
    const source = unpad(
      `
      function c() {
        let a = 10;
        const d = Number.isNaN(a);
        Math.max(a, b) + Math.max(b, a);
        return d && Number.isFinite(a);
      }
    `
    );
    // Jest arranges in alphabetical order, So keeping it as _source
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("should minify standard built in properties", () => {
    const source = unpad(
      `
      function a () {
        Number.NAN + Number.NAN;
        return Math.PI + Math.PI + Number.EPSILON + Number.NAN;
      }
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("should take no of occurences in to account", () => {
    const source = unpad(
      `
      function a() {
        Math.floor(a) + Math.floor(b) + Math.min(a, b);
      }
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("should collect and minify no matter any depth", () => {
    const source = unpad(
      `
      function a (){
        Math.max(b, a);
        function b() {
          const a = Math.floor(c);
          Math.min(b, a) * Math.floor(b);
          function c() {
            Math.floor(c) + Math.min(b, a)
          }
        }
      }
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("shouldn't minify builtins in the program scope to avoid leaking", () => {
    const source = unpad(
      `
      Math.max(c, d)
      function a (){
        Math.max(b, a) + Math.max(c, d);
      }
      Math.max(e, f)
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("should collect and minify in segments if there is no common ancestor", () => {
    const source = unpad(
      `
      function a(){
        function d(){
          Math.floor(as, bb);
        }
      }
      function b(){
        Math.floor(as, bb);
        function d(){
          Math.floor(as, bb);
        }
      }
      function c(){
        Math.floor(as, bb);
        function d(){
          Math.floor(as, bb);
        }
      }
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("should evalaute expressions if applicable and optimize it", () => {
    const source = unpad(
      `
      const a = Math.max(Math.floor(2), 5);
      let b = 1.8;
      let x = Math.floor(Math.max(a, b));
      foo(x);
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("should not evaluate if its side effecty", () => {
    const source = unpad(
      `
      Math.max(foo(), 1);
      Math.random();
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });

  it("should not minify for computed properties", () => {
    const source = unpad(
      `
      let max = "floor";
      Math[max](1.5);
    `
    );
    expect({ _source: source, expected: transform(source) }).toMatchSnapshot();
  });
});
