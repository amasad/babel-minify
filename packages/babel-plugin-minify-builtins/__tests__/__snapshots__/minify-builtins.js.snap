// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minify-builtins should collect and minify in segments if there is no common ancestor 1`] = `
Object {
  "_source": "function a(){
  function d(){
    Math.floor(as, bb);
  }
}
function b(){
  Math.floor(as, bb);
  function d(){
    Math.floor(as, bb);
  }
}
function c(){
  Math.floor(as, bb);
  function d(){
    Math.floor(as, bb);
  }
}",
  "expected": "function a() {
  function d() {
    Math.floor(as, bb);
  }
}
function b() {
  var _Mathfloor = Math.floor;

  _Mathfloor(as, bb);
  function d() {
    _Mathfloor(as, bb);
  }
}
function c() {
  var _Mathfloor2 = Math.floor;

  _Mathfloor2(as, bb);
  function d() {
    _Mathfloor2(as, bb);
  }
}",
}
`;

exports[`minify-builtins should collect and minify no matter any depth 1`] = `
Object {
  "_source": "function a (){
  Math.max(b, a);
  function b() {
    const a = Math.floor(c);
    Math.min(b, a) * Math.floor(b);
    function c() {
      Math.floor(c) + Math.min(b, a)
    }
  }
}",
  "expected": "function a() {
  Math.max(b, a);
  function b() {
    var _Mathmin = Math.min;
    var _Mathfloor = Math.floor;

    const a = _Mathfloor(c);
    _Mathmin(b, a) * _Mathfloor(b);
    function c() {
      _Mathfloor(c) + _Mathmin(b, a);
    }
  }
}",
}
`;

exports[`minify-builtins should evalaute expressions if applicable and optimize it 1`] = `
Object {
  "_source": "const a = Math.max(Math.floor(2), 5);
let b = 1.8;
let x = Math.floor(Math.max(a, b));
foo(x);",
  "expected": "const a = 5;
let b = 1.8;
let x = 5;
foo(x);",
}
`;

exports[`minify-builtins should minify standard built in methods 1`] = `
Object {
  "_source": "function c() {
  let a = 10;
  const d = Number.isNaN(a);
  Math.max(a, b) + Math.max(b, a);
  return d && Number.isFinite(a);
}",
  "expected": "function c() {
  var _Mathmax = Math.max;

  let a = 10;
  const d = false;
  _Mathmax(a, b) + _Mathmax(b, a);
  return d && true;
}",
}
`;

exports[`minify-builtins should minify standard built in properties 1`] = `
Object {
  "_source": "function a () {
  Number.NAN + Number.NAN;
  return Math.PI + Math.PI + Number.EPSILON + Number.NAN;
}",
  "expected": "function a() {
  var _MathPI = Math.PI;
  var _NumberNAN = Number.NAN;

  _NumberNAN + _NumberNAN;
  return _MathPI + _MathPI + Number.EPSILON + _NumberNAN;
}",
}
`;

exports[`minify-builtins should not evaluate if its side effecty 1`] = `
Object {
  "_source": "Math.max(foo(), 1);
Math.random();",
  "expected": "Math.max(foo(), 1);
Math.random();",
}
`;

exports[`minify-builtins should not minify for computed properties 1`] = `
Object {
  "_source": "let max = \\"floor\\";
Math[max](1.5);",
  "expected": "let max = \\"floor\\";
Math[max](1.5);",
}
`;

exports[`minify-builtins should take no of occurences in to account 1`] = `
Object {
  "_source": "function a() {
  Math.floor(a) + Math.floor(b) + Math.min(a, b);
}",
  "expected": "function a() {
  var _Mathfloor = Math.floor;

  _Mathfloor(a) + _Mathfloor(b) + Math.min(a, b);
}",
}
`;

exports[`minify-builtins shouldn't minify builtins in the program scope to avoid leaking 1`] = `
Object {
  "_source": "Math.max(c, d)
function a (){
  Math.max(b, a) + Math.max(c, d);
}
Math.max(e, f)",
  "expected": "Math.max(c, d);
function a() {
  var _Mathmax = Math.max;

  _Mathmax(b, a) + _Mathmax(c, d);
}
Math.max(e, f);",
}
`;
